writer.go

```diff
@@ -1,66 +1,214 @@
 package log
 
 import (
+	"fmt"
 	"io"
 	"sync"
-	"time"
 )
 
+var MaxWriterBuffer = 512 * 1024
+
+var log = Logger("eventlog")
+
 type MirrorWriter struct {
-	writers []io.WriteCloser
-	lk      sync.Mutex
+	active   bool
+	activelk sync.Mutex
+
+	// channel for incoming writers
+	writerAdd chan io.WriteCloser
+
+	// slices of writer/sync-channel pairs
+	writers []*bufWriter
+
+	// synchronization channel for incoming writes
+	msgSync chan []byte
+}
+
+type writerSync struct {
+	w  io.WriteCloser
+	br chan []byte
+}
+
+func NewMirrorWriter() *MirrorWriter {
+	mw := &MirrorWriter{
+		msgSync:   make(chan []byte, 64), // sufficiently large buffer to avoid callers waiting
+		writerAdd: make(chan io.WriteCloser),
+	}
+
+	go mw.logRoutine()
+
+	return mw
 }
 
 func (mw *MirrorWriter) Write(b []byte) (int, error) {
-	mw.lk.Lock()
-	// write to all writers, and nil out the broken ones.
-	var dropped bool
-	done := make(chan error, 1)
-	for i, w := range mw.writers {
-		go func(out chan error) {
-			_, err := w.Write(b)
-			out <- err
-		}(done)
+	mw.msgSync <- b
+	return len(b), nil
+}
+
+func (mw *MirrorWriter) logRoutine() {
+	for {
 		select {
-		case err := <-done:
-			if err != nil {
-				mw.writers[i].Close()
-				mw.writers[i] = nil
-				dropped = true
+		case b := <-mw.msgSync:
+			// write to all writers
+			dropped := mw.broadcastMessage(b)
+
+			// consolidate the slice
+			if dropped {
+				mw.clearDeadWriters()
 			}
-		case <-time.After(time.Millisecond * 500):
-			mw.writers[i].Close()
+		case w := <-mw.writerAdd:
+			mw.writers = append(mw.writers, newBufWriter(w))
+
+			mw.activelk.Lock()
+			mw.active = true
+			mw.activelk.Unlock()
+		}
+	}
+}
+
+// broadcastMessage sends the given message to every writer
+// if any writer is killed during the send, 'true' is returned
+func (mw *MirrorWriter) broadcastMessage(b []byte) bool {
+	var dropped bool
+	for i, w := range mw.writers {
+		_, err := w.Write(b)
+		if err != nil {
 			mw.writers[i] = nil
 			dropped = true
-
-			// clear channel out
-			done = make(chan error, 1)
 		}
 	}
+	return dropped
+}
 
-	// consolidate the slice
-	if dropped {
-		writers := mw.writers
-		mw.writers = nil
-		for _, w := range writers {
-			if w != nil {
-				mw.writers = append(mw.writers, w)
-			}
+func (mw *MirrorWriter) clearDeadWriters() {
+	writers := mw.writers
+	mw.writers = nil
+	for _, w := range writers {
+		if w != nil {
+			mw.writers = append(mw.writers, w)
 		}
 	}
-	mw.lk.Unlock()
-	return len(b), nil
+	if len(mw.writers) == 0 {
+		mw.activelk.Lock()
+		mw.active = false
+		mw.activelk.Unlock()
+	}
 }
 
 func (mw *MirrorWriter) AddWriter(w io.WriteCloser) {
-	mw.lk.Lock()
-	mw.writers = append(mw.writers, w)
-	mw.lk.Unlock()
+	mw.writerAdd <- w
 }
 
 func (mw *MirrorWriter) Active() (active bool) {
-	mw.lk.Lock()
-	active = len(mw.writers) > 0
-	mw.lk.Unlock()
+	mw.activelk.Lock()
```

Comment:
To allow things that would be logging to skip a costly string formatting operation if there are no log consumers.

overall operation is quite a bit slower when logs are being read due to the amount of base58 encoding going on.